pkg/controllers/user/windows/workload_test.go

```diff
@@ -0,0 +1,341 @@
+package windows
+
+import (
+	"testing"
+
+	util "github.com/rancher/rancher/pkg/controllers/user/workload"
+	"github.com/stretchr/testify/assert"
+	"k8s.io/api/apps/v1beta2"
+	batchv1 "k8s.io/api/batch/v1"
+	batchv1beta1 "k8s.io/api/batch/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+)
+
+func Test_canDeploy(t *testing.T) {
+	type testcase struct {
+		target    *util.Workload
+		canDeploy bool
+	}
+	cases := []testcase{
+		testcase{
+			target: &util.Workload{
+				TemplateSpec: &v1.PodTemplateSpec{
+					Spec: v1.PodSpec{
+						NodeSelector: map[string]string{"beta.kubernetes.io/os": "linux"},
+					},
+				},
+			},
+			canDeploy: true,
+		},
+		testcase{
+			target: &util.Workload{
+				TemplateSpec: &v1.PodTemplateSpec{
+					Spec: v1.PodSpec{
+						Affinity: &v1.Affinity{
+							NodeAffinity: &v1.NodeAffinity{
+								RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+									NodeSelectorTerms: []v1.NodeSelectorTerm{
+										v1.NodeSelectorTerm{
+											MatchExpressions: []v1.NodeSelectorRequirement{
+												v1.NodeSelectorRequirement{
+													Key:      "beta.kubernetes.io/os",
+													Operator: v1.NodeSelectorOpIn,
+													Values:   []string{"linux"},
+												},
+												v1.NodeSelectorRequirement{
+													Key:      "not-related",
+													Operator: v1.NodeSelectorOpIn,
+													Values:   []string{"not-related"},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			canDeploy: true,
+		},
+		testcase{
+			target: &util.Workload{
+				TemplateSpec: &v1.PodTemplateSpec{
+					Spec: v1.PodSpec{
+						Affinity: &v1.Affinity{
+							NodeAffinity: &v1.NodeAffinity{
+								RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+									NodeSelectorTerms: []v1.NodeSelectorTerm{
+										v1.NodeSelectorTerm{
+											MatchExpressions: []v1.NodeSelectorRequirement{
+												v1.NodeSelectorRequirement{
+													Key:      "not-related",
+													Operator: v1.NodeSelectorOpIn,
+													Values:   []string{"not-related"},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			canDeploy: false,
+		},
+		testcase{
+			target: &util.Workload{
+				TemplateSpec: &v1.PodTemplateSpec{
+					Spec: v1.PodSpec{},
+				},
+			},
+			canDeploy: false,
+		},
+	}
+
+	for _, c := range cases {
+		assert.Equal(t, canDeployedIntoLinuxNode(c.target), c.canDeploy, "failed to tell which workload wants to be deployed into linux")
+	}
+}
+
+type workloadRefenece struct {
+	workload *util.Workload
+	deploy   *v1beta2.Deployment
+	rc       *v1.ReplicationController
+	rs       *v1beta2.ReplicaSet
+	ss       *v1beta2.StatefulSet
+	ds       *v1beta2.DaemonSet
+	job      *batchv1.Job
+	cj       *batchv1beta1.CronJob
+}
+
+type workloadTestCase struct {
```

Comment:
Add `Name` to testcase, and print it on failure. It would help to know what each case is for, and to track test failure. The same to the other tests.